### PR TITLE
Blacklisting And Invalidating Sessions

### PR DESCRIPTION
**Pull request by nadnoslen**
Some global notes about this PR:
* Lots of tests in this PR.  Pull down the code and run.  I'd love to try this stuff against your ReactApp.
* `AuthenticationsController` is responsible for logging a user in with their email and password and also deleting their session by logging out
* `RenewalsController` is responsible for attempting to refresh an access token
* `api/v1/protected/SessionsController` is responsible for READONLY access to session information.  The refresh token id (`RUID`) is never transmitted.  There is one PATCH/PUT action; `invalidate`.  Passing a `SESSION.ID` (not `RUID`) to the `invalidate` method will attempt to invalidate that session by immediately expiring that refresh token thereby destroying that session's ability to _renew_ the next time the access token expires.
* `api/v1/protected/UsersController` is responsible for User CRUD

Lots of TODOs.  I'll markup #1 and this PR.

### CODE REVIEW

File: app/controllers/api/v1/protected/sessions_controller.rb [hunk 1]
```diff
@@ -0,0 +1,19 @@
+class Api::V1::Protected::SessionsController < ApplicationController
```
Comment (nadnoslen) [inline app/controllers/api/v1/protected/sessions_controller.rb]: Notice the path to this URL endpoint: `api/v1/protected/...`.  I envision some cases where we might expose information via a public endpoint `api/v1/public/...`.

File: app/controllers/authentications_controller.rb [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+    end
+  end
+
+  # logout
+  def destroy
+    session = JWTSessions::Session.new(payload: payload)
+    session.flush_by_access_payload
+
+    response.delete_cookie JWTSessions.access_cookie
+    # TODO update SessionsUser
```
Comment (nadnoslen) [inline app/controllers/authentications_controller.rb]: We need to use the RUID in the payload to lookup the appropriate `Session` and mark it as `invalidated: true` as well as set the `current_user` to the `invalidated_by`.

File: app/resources/api/v1/protected/session_resource.rb [hunk 3]
```diff
@@ -0,0 +1,28 @@
+class Api::V1::Protected::SessionResource < JSONAPI::Resource
```
Comment (nadnoslen) [inline app/resources/api/v1/protected/session_resource.rb]: The `JSONAPI::Resource` is what you call a DTO.  You create a resource for all endpoints.  For example, if we were going to grant public access to this resource we'd have to create it and a controller under the `api/v1/public/...` path.

File: app/resources/api/v1/protected/session_resource.rb [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+  immutable # no CUD through controller
```
Comment (nadnoslen) [inline app/resources/api/v1/protected/session_resource.rb]: This Resource/DTO is marked as immutable.  It cannot be altered through the controller or by api calls.

File: app/resources/api/v1/protected/session_resource.rb [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+      :browser_version,
+      :created_at,
+      :device,
+      :expiring_at,
+      :invalidated,
+      :ip_address,
+      :platform,
+      :platform_version,
+      :updated_at,
+      {}
```
Comment (nadnoslen) [inline app/resources/api/v1/protected/session_resource.rb]: RUID is never sent to the dirty world.

File: app/controllers/api/v1/protected/sessions_controller.rb (hunk 1 above)
Comment (bradf83) [reply app/controllers/api/v1/protected/sessions_controller.rb]: Sounds like a good idea to me

File: app/controllers/api/v1/protected/sessions_controller.rb [hunk 6]
```diff
@@ -0,0 +1,19 @@
...
+  before_action :authorize_access_request!
+
+  def invalidate
+    # TODO: protect this action using Pundit so that only Administrator role can do it
+    session = Session.find_by!(id: params[:id])
+
+    jwt_session = JWTSessions::Session.new(payload: payload)
+    # TODO: the following flush will throw an JWTSessions::Errors::Unauthorized exception if the refresh token
+    # cannot be found; what should we do?
+    jwt_session.flush_by_uid(session.ruid)
```
[the remark this answers is not in the document]
Comment (bradf83) [reply app/controllers/api/v1/protected/sessions_controller.rb]: How about we use a 422 Unprocessable Entity?

[the remark this answers is not in the document]
Comment (nadnoslen) [reply app/controllers/api/v1/protected/sessions_controller.rb]: Yeah.  That actually matches the pattern!

File: app/controllers/authentications_controller.rb [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+          browser: browser.name,
+          browser_version: browser.full_version,
+          device: browser.device.name,
+          expiring_at: Time.now, # TODO: must pull this from the refresh token
+          ip_address: request.remote_ip,
+          platform: browser.platform.name,
+          platform_version: browser.platform.version,
+          ruid: session.payload['ruid'],
+          user: user
+      )
```
[the remark this answers is not in the document]
Comment (bradf83) [reply app/controllers/authentications_controller.rb]: I think 15 minutes is a good time window especially for less critical information apps, in a critical information app I may say checking the database every time might be more acceptable.  I like the extra user agent information.

File: app/resources/api/v1/protected/session_resource.rb (hunk 3 above)
Comment (bradf83) [reply app/resources/api/v1/protected/session_resource.rb]: Makes sense.

Comment (nadnoslen) [reply app/resources/api/v1/protected/session_resource.rb]: The one problem I'm starting see is at the EmberJs/ReactApp level.  The data adapter down at that level will need to speak to two different endpoints.  That might get complicated.

File: app/controllers/renewals_controller.rb [hunk 8]
```diff
@@ -1,4 +1,4 @@
-class RefreshController < ApplicationController
+class RenewalsController < ApplicationController
```
Comment (bradf83) [inline app/controllers/renewals_controller.rb]: Not sure I like Renewals.

File: app/controllers/authentications_controller.rb (hunk 7 above)
[the remark this answers is not in the document]
Comment (nadnoslen) [reply app/controllers/authentications_controller.rb]: I will unit test what is going on in JWTSessions to understand exactly what is happening when we flush the different kinds of tokens.

File: app/controllers/renewals_controller.rb (hunk 8 above)
Comment (nadnoslen) [reply app/controllers/renewals_controller.rb]: That's nice!  Tell me what to call it!  Controllers should be pluralized.  What's the plural to "Refresh"?

File: app/controllers/api/v1/protected/sessions_controller.rb (hunk 1 above)
Comment (nadnoslen) [reply app/controllers/api/v1/protected/sessions_controller.rb]: Might have to refactor this for EmberJS!  I'll test it out soon.

**Pull request merged**